src/xdp-utils.c

```diff
@@ -2447,6 +2447,129 @@ xdp_validate_serialized_icon (GVariant  *v,
   return TRUE;
 }
 
+#define SOUND_VALIDATOR_GROUP "Sound Validator"
+
+gboolean
+xdp_validate_serialized_sound (GVariant  *v)
+{
+  const char *key;
+  g_autoptr(GVariant) value = NULL;
+  __attribute__((cleanup(cleanup_temp_file))) char *tmp_sound_path = NULL;
+  g_autofree char *file_sound_path = NULL;
+  const char *sound_path;
+  xdp_autofd int fd = -1;
+  g_autoptr(GOutputStream) stream = NULL;
+  int status;
+  g_autofree char *format = NULL;
+  g_autofree char *stdoutlog = NULL;
+  g_autofree char *stderrlog = NULL;
+  g_autoptr(GError) error = NULL;
+  const char *sound_validator = LIBEXECDIR "/xdg-desktop-portal-validate-sound";
+  const char *args[6];
+  /* same allowed formats as Flatpak */
+  const char *allowed_sound_formats[] = { "ogg/opus", "ogg/vorbis", "wav/pcm", NULL };
+  g_autoptr(GKeyFile) key_file = NULL;
+
+  if (g_getenv ("XDP_VALIDATE_SOUND"))
+    sound_validator = g_getenv ("XDP_VALIDATE_SOUND");
+
+  if (!g_file_test (sound_validator, G_FILE_TEST_EXISTS))
+    {
+      g_warning ("Sound validation: %s not found, rejecting sound by default.", sound_validator);
+      return FALSE;
+    }
+
+  g_variant_get (v, "(&sv)", &key, &value);
+
+  if (strcmp (key, "bytes") == 0 && g_variant_is_of_type (value, G_VARIANT_TYPE_BYTESTRING))
+    {
+      g_autoptr(GBytes) bytes;
+      gconstpointer bytes_data;
+      gsize bytes_len;
+
+      bytes = g_variant_get_data_as_bytes (value);
+
+      fd = g_file_open_tmp ("soundXXXXXX", &tmp_sound_path, &error);
+      if (fd == -1)
+        {
+          g_warning ("Sound validation: %s", error->message);
+          return FALSE;
+        }
+
+      stream = g_unix_output_stream_new (fd, FALSE);
+
+      /* Use write_all() instead of write_bytes() so we don't have to worry about
+       * partial writes (https://gitlab.gnome.org/GNOME/glib/-/issues/570).
+       */
+      bytes_data = g_bytes_get_data (bytes, &bytes_len);
+      if (!g_output_stream_write_all (stream, bytes_data, bytes_len, NULL, NULL, &error))
+        {
+          g_warning ("Sound validation: %s", error->message);
+          return FALSE;
+        }
+
+      if (!g_output_stream_close (stream, NULL, &error))
+        {
+          g_warning ("Sound validation: %s", error->message);
+          return FALSE;
+        }
+
+      sound_path = tmp_sound_path;
+    }
+  else if (strcmp (key, "file") == 0 && g_variant_is_of_type (value, G_VARIANT_TYPE_STRING))
+    {
+      file_sound_path = g_filename_from_uri (g_variant_get_string (value, NULL), NULL, &error);
```

Comment:
the file is opened in a sandbox, same as the image_validater.

Reply:
Yes, it is opened in a sandbox but the path is app controlled, so it can make the portal open files in e.g. `~/.ssh`. The validation program should in theory notice that something isn't a sound file.

Ignoring the security implications, the reason why this is still bad is because the app doesn't know the host mount ns and where files are. This really needs to only take files from the sandbox, and the easiest way to ensure that is by requiring the app to send a fd because that's proof that the app can access the file.

Reply:
No this path isn't created by the sandbox. The app gives a fd to the portal frontend and converts it to a file path (using `xdp_app_info_get_path_for_fd ()`). Then it gives the file path to the backend.

Reply:
ah, right, already forgot about that part. all good then.